Modifications to the NMR schema package

Sathya-S3 opened the pull request:
Hi @andreaa93, @jkshenton, @JFRudzinski,

This pull request introduces significant updates and refinements to the NMR schema within the repository. The changes aim to improve clarity, accuracy, and alignment with the .magres file structure, ensuring the schema better represents the data and metadata fields used by the NMR crystallography community. These updates follow quantity definitions described in the magres specification (https://www.ccpnc.ac.uk/docs/magres/magres-format.pdf) and other scientific literature, all tagged in quantity descriptions where appropriate.

**Note:**
I would like to welcome @jryates and @dceresoli for comments on a few magres file related entries and quantities output by CASTEP and QE-GIPAW calculations. Could you kindly provide access to them for observing and co-reviewing this PR? I have a dedicated questions section at the bottom of the PR notes, for them.

Modifications to the schema package:
**Magnetic Shielding:**
1. **Magnetic Shielding:** Renamed class **MagneticShieldingTensor** to **MagneticShielding** for consistency with the naming conventions used across the schema. The functionality and logic of the class remain unchanged.

**Electric Field Gradient:**
1. **Decomposition of Electric Field Gradient contributions**: To preserve the decomposed EFG contributions reported in some DFT codes, I have separated the **ElectricFieldGradient** class into three separate classes: The default **ElectricFieldGradient** class always receives the total contribution as EFG tensors per atomic site, the new **ElectricFieldGradientLocal** (local contribution), and **ElectricFieldGradientNonlocal** (non-local contribution) classes have the same data type and shape as the main class, but are in place to store the decomposed contributions, if present in the uploaded files. This change also aligns with the quantities permissible in .magres file structure and improves clarity.
2. **Updated the formula and description** for **quadrupolar_coupling_constant** in ElectricFieldGradient class to ensure correctness and provide clearer explanation.
3. **Removed the ElectricFieldGradients class** in favour of defining total, local, and non-local components directly in the **Outputs** class as subsections. This is also consistent with @andreaa93's advice that the "Output class will contain a flat list of all physical properties without having some of them that act only as wrapper for more physical properties".
4. Commented out the **type** field in the ElectricFieldGradient class, as its functionality is now handled by the separate classes for local and non-local contributions.
5. Integrated all contributions into the **Outputs** class as subsections, ensuring proper representation in the schema.

**Indirect Spin-Spin Coupling:**
1. Renamed the **SpinSpinCoupling** class to **IndirectSpinSpinCoupling** for accuracy. The `isc` tag in .magres file represents the reduced spin coupling tensor (indirect spin coupling).
2. Introduced the **jcoupling_value** field in the **IndirectSpinSpinCoupling** class to represent the J-coupling tensor. This field is derived from the indirect spin-spin coupling tensor as a scaled value using gyromagnetic ratios and Planck's constant.
3. Decomposed the **IndirectSpinSpinCoupling** class into separate classes for each contribution:

    -  **IndirectSpinSpinCouplingFermiContact** (`isc_fc` tag in the .magres file)
    - **IndirectSpinSpinCouplingOrbitalDiamagnetic** (`isc_orbital_d` tag in the .magres file)
    - **IndirectSpinSpinCouplingOrbitalParamagnetic** (`isc_orbital_p` tag in the .magres file)
    - **IndirectSpinSpinCouplingSpinDipolar** (`isc_spin` tag in the .magres file)
   These classes have the same data type and tensor shape as the main **IndirectSpinSpinCoupling** class, but are in place to store the decomposed contributions, if present in the uploaded files. This change also aligns with the quantities permissible in .magres file structure and improves clarity.
4. Commented out the **type** field in the **IndirectSpinSpinCoupling** class, as its functionality is now handled by the separate classes for each contribution.
5. Integrated all contributions into the **Outputs** class as subsections, ensuring proper representation in the schema.
    
**Magnetic Susceptibility:**
1. Refined the **MagneticSusceptibility** class to focus solely on the macroscopic magnetic susceptibility tensor:
     - Removed the **scale_dimension** field, which previously distinguished between "microscopic" and "macroscopic" tensors.
     - Updated the **value** field to include a unit of 10 ** -6 * cm ** 3 / mol, which is consistent with units prescribed in .magres.
2. Updated citation/reference in description. Clarified that the tensor is identified by the 'sus' tag in the .magres file.

Questions for @jryates, @dceresoli:
1. Besides `efg_local` and `efg_nonlocal`, are there any other decomposed contributions to `efg` that are to be expected in magres files or computed in CASTEP or QE?
2. For the sake of computing j-coupling tensor, are the gyromagnetic ratios potentially provided with the calculation metadata in CASTEP or QE or somewhere in the magres file? Is there usefulness to pre-computing and storing this tensor in the database for searchability, or is it more usable to compute on the fly when visualising parameters at a higher level in the frontend using structure visualisation tools like MagresView2.0 that can already visualise magnetic shielding and electric field gradient values per atomic site.
3. For magnetic susceptibility, DFT codes like VASP and QE seem to compute one tensor with `pGv-approximation`, another with `vGv-approximation`. QE, however, writes the average of these 2 tensors into a .magres file. Where a user uploads both the computation files and also a magres file that was output from the computation process, I would like to capture all available information. What is your opinion on introducing tags to identify each different magnetic susceptibility data? What tags should be introduced?

To Do List:
- [x] Normalising code to resolve diagonalised (eigen) values from the EFG tensor and consequently the asymmetry parameter needs to be added (commits: [d74393b](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/d74393b15b212d742451e48d3b1b76fc33d4ac70), [9f6e106](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/9f6e1064d51eaf8d9bd75bbb02343d899a4702c6)).
- [x] Normalising code to resolve diagonalised (eigen) values from the Magnetic Shielding tensor and consequently the related useful parameters such as the span and skew of the tensor (isotropic value is already extracted inside the class) (commits: [d74393b](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/d74393b15b212d742451e48d3b1b76fc33d4ac70), [1366f02](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/1366f02a120578d0d48ef481afbb4d6190f0f16a)).
- [x] Normalising code to resolve diagonalised (eigen) values from the Indirect Spin-Spin Coupling (reduced coupling) tensor and consequently the related useful parameters (commits: [d74393b](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/d74393b15b212d742451e48d3b1b76fc33d4ac70), [fc0a441](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/fc0a441d9ca392d60d3f8a46b3a9e608e23cac90)).
- [x] Update magnetic susceptibility class with appropriate calculation approximation tags (commit: [eaeb789](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/eaeb789f1a9977986c97fa13cf9924f1dbf75440)).
- [x] Fix the 'AttributeError: Name' when the magnetic susceptibility class is parsed (commit: [5c17e57](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/5c17e5796dfe3b8a1f78bbd1da88860991c4a64f)).
- [x] Extend testing to all core Physical Quantities (at the moment, only magnetic susceptibility tensor is included)

Sathya-S3 committed:
make ElectricFieldGradient class explicit for total efg

Sathya-S3 committed:
correct definition for quadrupolar_coupling_constant

Sathya-S3 committed:
add classes for local and nonlocal efg

Sathya-S3 committed:
commented out nested efg contribution subsections

Sathya-S3 committed:
linting changes to quotation marks

Sathya-S3 committed:
efg description edit

Sathya-S3 committed:
redefine indirect spin-spin coupling class

Sathya-S3 committed:
individual classes for indirect spin coupling contributions

Sathya-S3 committed:
removed 'Tensor' from class name

Sathya-S3 committed:
changes to magnetic susceptitiblity class

Sathya-S3 committed:
mag sus class description edit

Sathya-S3 committed:
subsection variables for decomposed contributions in Outputs

Sathya-S3 committed:
magnetic susceptibility class edits

Sathya-S3 committed:
mag sus correction - one tensor per crystal

Sathya-S3 committed:
mag sus - added field to accommodate approximation tags

Sathya-S3 committed:
simplified mag sus value definition

Sathya-S3 committed:
update test.archive.yaml value

Sathya-S3 committed:
add self.name to Mag.Sus. class init

Sathya-S3 committed:
ruff linting fix

Sathya-S3 committed:
ruff formatting fix

Sathya-S3 committed:
resolve function for single tensor like magnetic susceptibility

Sathya-S3 committed:
ruff fornat for new function

Sathya-S3 commented:
@andreaa93 I may have misinterpreted the pytest error for 'AttributeError: name'. Could you help with the failing tests please?

I have tried to write a separate resolve name function for MagneticSusceptibility because it does not use the resolve function already written which assumes atom-resolved 'PhysicalProperty' and cycles through entities (atom sites) to extract tensors in a loop. Mag. Sus will just be one tensor for the entire crystal and I kind of tried to pass the class name as the resolved quantity name, but it doesn't quite get past the error. Thank you very much in advance.

Sathya-S3 commented:
@cecizag Hi Ceclia, this is the PR Kane mentioned in the meeting. Please tag all your colleagues here as you see fit.

Sathya-S3 committed:
single tensor resolve function deleted

Sathya-S3 committed:
Mag Sus - add shape to value, remove scaling factor from unit

Sathya-S3 committed:
update tests

Sathya-S3 committed:
remove quad coupling constant - not to be stored

dceresoli commented:
_Q: Besides efg_local and efg_nonlocal, are there any other decomposed contributions to efg that are to be expected in magres files or computed in CASTEP or QE?_
A: the NMR shielding can be decomposed as well, but to me, the corresponding tags should be optional.

_Q: For magnetic susceptibility, DFT codes like VASP and QE seem to compute one tensor with pGv-approximation, another with vGv-approximation. QE, however, writes the average of these 2 tensors into a .magres file. Where a user uploads both the computation files and also a magres file that was output from the computation process, I would like to capture all available information. What is your opinion on introducing tags to identify each different magnetic susceptibility data? What tags should be introduced?_
A: I'm in favor of introducing new tags.

Sathya-S3 committed:
eigenvalues - standard, haeberlen conventions

Sathya-S3 committed:
Magnetic shielding - eigenvalues and parameters

Sathya-S3 committed:
Electric Field Gradient - eigenvalues and parameters

Sathya-S3 committed:
remove quadrupolar coupling constant

Sathya-S3 committed:
Reduced Spin Coupling - Eigenvalues and parameters

Sathya-S3 committed:
ruff linting fix

Sathya-S3 committed:
Magnetic Susceptibility tags added

Sathya-S3 commented:
@Pepe-Marquez @andreaa93

I am done with all listed actions on the PR, except adding more test cases for the NMR quantities.

@jkshenton and I have a query that we want to raise about changing the bit in the schema which resolves the atomic site names to assign labels to stored tensor data. I have raised this in Discord within the issues 'Failing test in plugin' where we were already chatting about an issue in this PR. Could you please help?

**For context into our query:**
In src/nomad_nmr_schema/schema_packages/schema_package.py, within resolve_name_from_entity_ref() where the name is resolved for each atomic site from the input list of entities, the **_nomads_simulations.schema_packages.atoms_state.AtomsState_** is used to read chemical symbols and optionally, attach an index field to the chemical symbol to uniquely identify the atom within the material. I understand this is to stay in line with some magres file examples to label each atomic site as hydrogen atoms 'H,1', 'H,2', etc. and carbon atoms 'C,1', 'C,2', etc. and so on.

But, recently we realised the possibility that in some DFT codes, you can assign custom site labels based on atomic symmetry. If a few H atoms correspond to a certain symmetry there could be four different H atoms labelled 'H1' or 'H1a' (user-chosen labels) which can be further resolved to unique sites as 'H1,1', 'H1,2', etc. So I need a way to record these custom labels when storing the tensors plus indices. I noticed that another quantity **_nomad.datamodel.metainfo.aystem.Atoms.labels_** from the metainfo library may be suitable for this. Am I allowed to use this or was this reserved just for the older data schema?

**Follow-up question:** If working with custom data labels such as H1, H1a, etc, would it still be possible to extract the unique elements from this list to record?

andreaa93 reviewed (approved):
Everything looks good to me. I think you can use the nomad.datamodel.metainfo.aystem.Atoms.labels quantity as you described!

Cheers

Sathya-S3 committed:
Remove efg local and nonlocal classes

Sathya-S3 committed:
remove efg local and nonlocal from Outputs

Commit:
Add NMRTensor class and associated tensor utility functions

- Implemented NMRTensor class for handling NMR tensor operations, including tensor diagonalization and eigenvalue sorting.
- Introduced TensorConvention enum for sorting eigenvalues in different conventions (Haeberlen, Increasing, Decreasing, NQR).
- Added utility functions for calculating anisotropy, asymmetry, span, and skew of tensors.
- Created unit tests for NMRTensor initialization, eigenvalue ordering, and tensor properties to ensure correctness and reliability.
- Refactored the schema entries to use the NMRTensor class
- Refactored indirect spin spin coupling classes using a base class covering all.
- Made field naming consistent (isotropy, anisotopy etc)
- Changes EFG units to au - though I'm not sure if Nomad will accept that.

Commit:
Linting fixes

Commit:
ruff formatting

jkshenton commented:
Hi all,

So I've just pushed some changes to try to make for some cleaner code - by reusing a common NMRTensor class that does the processing of the derived quantities from the raw tensor data. 

- Implemented NMRTensor class for handling NMR tensor operations, including tensor diagonalization and eigenvalue sorting.
- Introduced TensorConvention enum for sorting eigenvalues in different conventions (Haeberlen, Increasing, Decreasing, NQR).
- Added utility functions for calculating anisotropy, asymmetry, span, and skew of tensors.
- Created unit tests for NMRTensor initialization, eigenvalue ordering, and tensor properties to ensure correctness and reliability.
- Refactored the schema entries to use the NMRTensor class
- Refactored indirect spin spin coupling classes using a base class covering the total ISC as well as the contributions.
- Made field naming consistent (isotropy, anisotopy etc)
- Changed EFG units to `au`. This one might benefit from some discussion - does NOMAD accept this unit? Is it the most appropriate as the standard unit for Nomad? Any thoughts on that @jryates and @dceresoli ?

I think we're nearly there with this PR. The only thing left of @Sathya-S3 's todo list is extend the testing, which he will tackle next

Sathya-S3 committed:
Removed Soprano copyright message

Sathya-S3 committed:
Corrected tensor ordering for EFG

dceresoli commented:
>   * Changed EFG units to |au|. This one might benefit from some discussion -
>     does NOMAD accept this unit? Is it the most appropriate as the standard unit
>     for Nomad? Any thoughts on that @jryates <https://github.com/jryates> and
>     @dceresoli <https://github.com/dceresoli> ?
I'm fine with |au| units, even though the docs should specify it's
"Hartree atomic units", not "Rydberg atomic units"

D.

Sathya-S3 committed:
nomad-simulations version upgrade

Sathya-S3 committed:
parse index with chemical symbol for label

Sathya-S3 committed:
Magres quantities test - without site label

Sathya-S3 committed:
change back to old code for time being

Sathya-S3 committed:
entity_ref corrected type in isc

Sathya-S3 committed:
Updated tests - check tensors, derived quantitites, name resolution

Sathya-S3 commented:
Hi @jkshenton , @andreaa93 ,

As of my latest commit [077c321](https://github.com/FAIRmat-NFDI/nomad-schema-plugin-nmr/pull/1/commits/077c32198988ac348c355e62a4c9edb4e9cc509f), I have added a comprehensive list of tests.

The tests include:
1. Basic tensor checks to make sure the normalize functions return the expected tensor after parsing.
2. Derived quantities checks calculated by the NMRTensor class, against manually (or separately programmatically) calculated values.
3. Name resolution checks for site labels containing the correct chemical symbol in the case of quantities magnetic shielding and electric field gradient, correct chemical symbol pairs in the case of indirect spin-spin coupling.

With reference to updates in #12 , I can now change the code in resolve_name_from_entity_ref() to include custom site labels and update the tests accordingly. Once this small step is complete, the PR can be closed.

Sathya-S3 committed:
Adopted 'label' from AtomsState for custom site labels

Sathya-S3 committed:
Included indices to site labels

Sathya-S3 commented:
Hello @EBB2675, @jkshenton 

In the last 2 commits, I have replaced the use of AtomsState.chemical_symbol with AtomsState.label for resolving custom site labels as site names on NOMAD.

In my test scripts, wherever parser function needs to be mimicked to pass information to the name resolving function, I have replaced AtomicCell.atoms_state references with ModelSystem.particle_states (as advised in #12) as the parent for AtomsState object that is assigned to entity_ref for resolving.

I think this is one of the last actions needed before closing the PR, subject to one final check.

Sathya-S3 committed:
Removed indices from custom site name labels

Sathya-S3 commented:
Thank you to @andreaa93, @jkshenton, @dceresoli @EBB2675, who have helped us review and improve this schema package to completion. All agreed actions on the to-do list have now been completed and the PR is ready to be closed now. I'll merge it with the main branch.

Sathya-S3 committed:
added clarification on efg units
